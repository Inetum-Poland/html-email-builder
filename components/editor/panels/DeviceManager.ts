import type { Editor, PanelProperties } from "grapesjs";
import { Monitor, Tablet, TabletSmartphone } from "lucide-static";
import { Device } from "@types";

const setDevice = (editor: Editor, device: Device) => {
  editor.setDevice(device);

  if (device !== Device.Desktop) {
    document.querySelector("#set-device-desktop")?.classList.remove("gjs-pn-active");
  }
};

const resetDevice = (editor: Editor) => {
  editor.setDevice(Device.Desktop);
  document.querySelector("#set-device-desktop")?.classList.add("gjs-pn-active");
};

const toggle = (device: Device) => ({
  run: (editor: Editor) => setDevice(editor, device),
  stop: resetDevice,
});

export const DeviceManager = (): PanelProperties => {
  const { $i18n: { t } } = useNuxtApp();

  return {
    id: "device-manager",
    buttons: [
      {
        active: true,
        attributes: { id: "set-device-desktop", title: t("desktop") },
        command: toggle(Device.Desktop),
        id: "set-device-desktop",
        label: Monitor,
        tagName: "button",
        togglable: false,
      },
      {
        attributes: { title: t("tablet") },
        command: toggle(Device.Tablet),
        id: "set-device-tablet",
        label: Tablet,
        tagName: "button",
      },
      {
        attributes: { title: t("mobile") },
        command: toggle(Device.Mobile),
        id: "set-device-mobile",
        label: TabletSmartphone,
        tagName: "button",
      },
    ],
  };
};
