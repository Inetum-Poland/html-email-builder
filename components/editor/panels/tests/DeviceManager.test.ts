import { expect, test } from "vitest";
import { DeviceManager } from "../DeviceManager";

test("renders the Device Manager panel", () => {
  expect(DeviceManager()).toMatchSnapshot();
});
