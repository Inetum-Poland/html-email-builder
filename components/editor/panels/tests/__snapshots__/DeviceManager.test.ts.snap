// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders the Device Manager panel 1`] = `
{
  "buttons": [
    {
      "active": true,
      "attributes": {
        "id": "set-device-desktop",
        "title": "Desktop",
      },
      "command": {
        "run": [Function],
        "stop": [Function],
      },
      "id": "set-device-desktop",
      "label": "
<svg
  class="lucide lucide-monitor"
  xmlns="http://www.w3.org/2000/svg"
  width="24"
  height="24"
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  stroke-width="2"
  stroke-linecap="round"
  stroke-linejoin="round"
>
  <rect width="20" height="14" x="2" y="3" rx="2" />
  <line x1="8" x2="16" y1="21" y2="21" />
  <line x1="12" x2="12" y1="17" y2="21" />
</svg>
",
      "tagName": "button",
      "togglable": false,
    },
    {
      "attributes": {
        "title": "Tablet",
      },
      "command": {
        "run": [Function],
        "stop": [Function],
      },
      "id": "set-device-tablet",
      "label": "
<svg
  class="lucide lucide-tablet"
  xmlns="http://www.w3.org/2000/svg"
  width="24"
  height="24"
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  stroke-width="2"
  stroke-linecap="round"
  stroke-linejoin="round"
>
  <rect width="16" height="20" x="4" y="2" rx="2" ry="2" />
  <line x1="12" x2="12.01" y1="18" y2="18" />
</svg>
",
      "tagName": "button",
    },
    {
      "attributes": {
        "title": "Mobile",
      },
      "command": {
        "run": [Function],
        "stop": [Function],
      },
      "id": "set-device-mobile",
      "label": "
<svg
  class="lucide lucide-tablet-smartphone"
  xmlns="http://www.w3.org/2000/svg"
  width="24"
  height="24"
  viewBox="0 0 24 24"
  fill="none"
  stroke="currentColor"
  stroke-width="2"
  stroke-linecap="round"
  stroke-linejoin="round"
>
  <rect width="10" height="14" x="3" y="8" rx="2" />
  <path d="M5 4a2 2 0 0 1 2-2h12a2 2 0 0 1 2 2v16a2 2 0 0 1-2 2h-2.4" />
  <path d="M8 18h.01" />
</svg>
",
      "tagName": "button",
    },
  ],
  "id": "device-manager",
}
`;
